From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tehin <62246450+Tehin@users.noreply.github.com>
Date: Tue, 17 Dec 2024 18:44:44 -0500
Subject: [PATCH] HoLA


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index ae4ebf509837e8d44255781c61d02873f8b74be8..369a546ff6e85b0f03b19f118be882191c37d607 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1270,6 +1270,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             }
             // Paper end - Add onboarding message for initial server start
 
+            LOGGER.info("HoLA");
+
             while (this.running) {
                 long i;
 
