From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tehin <62246450+Tehin@users.noreply.github.com>
Date: Tue, 17 Dec 2024 18:56:08 -0500
Subject: [PATCH] rollback


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 369a546ff6e85b0f03b19f118be882191c37d607..ae4ebf509837e8d44255781c61d02873f8b74be8 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1270,8 +1270,6 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             }
             // Paper end - Add onboarding message for initial server start
 
-            LOGGER.info("HoLA");
-
             while (this.running) {
                 long i;
 
